File: backend/comunidad/services/admin_panel.py
"""
Capa de Servicios — Panel de Administracion (Sprint 2 HU3).

Orquesta CRUD completo sobre todas las entidades del sistema.
Cada metodo valida que el usuario sea administrador antes de ejecutar.
"""
from ..negocio.admin import (
    validar_es_administrador,
    validar_puede_eliminar_usuario,
    validar_puede_cambiar_rol,
    validar_busqueda,
)
from ..interfaces.service_interfaces import AdminPanelInterface
from ..repositorios_implementacion import (
    UsuarioRepository,
    PublicacionRepository,
    TruequeRepository,
    ResenaRepository,
    TruequeMultipleRepository,
    SaldoComercialRepository,
    ResenaMultipleRepository,
)


class AdminPanelService(AdminPanelInterface):
    """Implementacion del servicio de administracion completo."""

    def __init__(
        self,
        usuario_repo=None,
        publicacion_repo=None,
        trueque_repo=None,
        resena_repo=None,
        trueque_multiple_repo=None,
        saldo_repo=None,
        resena_multiple_repo=None,
    ):
        self.usuario_repo = usuario_repo or UsuarioRepository()
        self.publicacion_repo = publicacion_repo or PublicacionRepository()
        self.trueque_repo = trueque_repo or TruequeRepository()
        self.resena_repo = resena_repo or ResenaRepository()
        self.trueque_multiple_repo = trueque_multiple_repo or TruequeMultipleRepository()
        self.saldo_repo = saldo_repo or SaldoComercialRepository()
        self.resena_multiple_repo = resena_multiple_repo or ResenaMultipleRepository()

    # ── Dashboard ─────────────────────────────────────────────────────────────

    def obtener_dashboard(self, admin):
        validar_es_administrador(admin)
        from backend.comunidad.models import (
            Publicacion, AcuerdoTrueque, Resena,
            AcuerdoTruequeMultiple, ResenaMultiple, SaldoComercial,
        )
        stats_usuarios = self.usuario_repo.contar_estadisticas()
        return {
            'usuarios': stats_usuarios,
            'publicaciones': {
                'total': Publicacion.objects.count(),
                'activas': Publicacion.objects.filter(esta_activa=True).count(),
            },
            'trueques': {
                'total': AcuerdoTrueque.objects.count(),
                'finalizados': AcuerdoTrueque.objects.filter(estado='FINALIZADO').count(),
                'pendientes': AcuerdoTrueque.objects.filter(estado='PENDIENTE').count(),
            },
            'trueques_multiples': {
                'total': AcuerdoTruequeMultiple.objects.count(),
                'finalizados': AcuerdoTruequeMultiple.objects.filter(estado='FINALIZADO').count(),
            },
            'resenas': {
                'total': Resena.objects.count(),
            },
            'resenas_multiples': {
                'total': ResenaMultiple.objects.count(),
            },
            'saldos': {
                'total_movimientos': SaldoComercial.objects.count(),
            },
        }

    # ── Usuarios ──────────────────────────────────────────────────────────────

    def listar_usuarios(self, admin, busqueda=None):
        validar_es_administrador(admin)
        termino = validar_busqueda(busqueda)
        return self.usuario_repo.listar_todos(termino)

    def toggle_usuario(self, admin, usuario_id):
        validar_es_administrador(admin)
        usuario = self.usuario_repo.obtener_por_id(usuario_id)
        if not usuario:
            raise ValueError('Usuario no encontrado.')
        return self.usuario_repo.actualizar_estado(usuario_id, not usuario.is_active)

    def cambiar_rol(self, admin, usuario_id, is_staff):
        validar_es_administrador(admin)
        validar_puede_cambiar_rol(admin)
        usuario = self.usuario_repo.obtener_por_id(usuario_id)
        if not usuario:
            raise ValueError('Usuario no encontrado.')
        return self.usuario_repo.actualizar_rol(usuario_id, is_staff)

    def eliminar_usuario(self, admin, usuario_id):
        validar_es_administrador(admin)
        validar_puede_eliminar_usuario(admin.id, usuario_id)
        usuario = self.usuario_repo.obtener_por_id(usuario_id)
        if not usuario:
            raise ValueError('Usuario no encontrado.')
        self.usuario_repo.eliminar(usuario_id)
        return {'eliminado': True}

    # ── Publicaciones ─────────────────────────────────────────────────────────

    def listar_publicaciones(self, admin, busqueda=None):
        validar_es_administrador(admin)
        termino = validar_busqueda(busqueda)
        return self.publicacion_repo.listar_todas(termino)

    def crear_publicacion_admin(self, admin, datos):
        validar_es_administrador(admin)
        if not datos.get('titulo'):
            raise ValueError('El titulo es requerido.')
        if not datos.get('descripcion'):
            raise ValueError('La descripcion es requerida.')
        if not datos.get('categoria'):
            raise ValueError('La categoria es requerida.')
        return self.publicacion_repo.crear(admin.id, {
            'tipo': datos.get('tipo', 'TALENTO'),
            'titulo': datos['titulo'],
            'descripcion': datos['descripcion'],
            'categoria': datos['categoria'],
            'urgencia': datos.get('urgencia', 'NORMAL'),
            'esta_activa': datos.get('esta_activa', True),
        })

    def moderar_publicacion(self, admin, publicacion_id, esta_activa):
        validar_es_administrador(admin)
        return self.publicacion_repo.actualizar_estado_admin(publicacion_id, esta_activa)

    def eliminar_publicacion(self, admin, publicacion_id):
        validar_es_administrador(admin)
        pub = self.publicacion_repo.obtener_por_id(publicacion_id)
        if not pub:
            raise ValueError('Publicacion no encontrada.')
        self.publicacion_repo.eliminar(publicacion_id)
        return {'eliminado': True}

    # ── Trueques ──────────────────────────────────────────────────────────────

    def listar_trueques(self, admin, busqueda=None):
        validar_es_administrador(admin)
        termino = validar_busqueda(busqueda)
        return self.trueque_repo.listar_todos(termino)

    def actualizar_estado_trueque(self, admin, trueque_id, estado):
        validar_es_administrador(admin)
        estados_validos = ['PENDIENTE', 'ACEPTADO', 'RECHAZADO', 'EN_CURSO', 'FINALIZADO']
        if estado not in estados_validos:
            raise ValueError(f'Estado invalido. Debe ser uno de: {", ".join(estados_validos)}')
        return self.trueque_repo.actualizar_estado_admin(trueque_id, estado)

    def eliminar_trueque(self, admin, trueque_id):
        validar_es_administrador(admin)
        trueque = self.trueque_repo.obtener_por_id(trueque_id)
        if not trueque:
            raise ValueError('Trueque no encontrado.')
        self.trueque_repo.eliminar(trueque_id)
        return {'eliminado': True}

    # ── Trueques Multiples ────────────────────────────────────────────────────

    def listar_trueques_multiples(self, admin, busqueda=None):
        validar_es_administrador(admin)
        termino = validar_busqueda(busqueda)
        return self.trueque_multiple_repo.listar_todos(termino)

    def actualizar_estado_trueque_multiple(self, admin, trueque_id, estado):
        validar_es_administrador(admin)
        estados_validos = ['PENDIENTE', 'ACEPTADO', 'RECHAZADO', 'EN_CURSO', 'FINALIZADO', 'EXPIRADO']
        if estado not in estados_validos:
            raise ValueError(f'Estado invalido. Debe ser uno de: {", ".join(estados_validos)}')
        return self.trueque_multiple_repo.actualizar_estado_admin(trueque_id, estado)

    def eliminar_trueque_multiple(self, admin, trueque_id):
        validar_es_administrador(admin)
        tm = self.trueque_multiple_repo.obtener_por_id(trueque_id)
        if not tm:
            raise ValueError('Trueque multiple no encontrado.')
        self.trueque_multiple_repo.eliminar(trueque_id)
        return {'eliminado': True}

    # ── Resenas ───────────────────────────────────────────────────────────────

    def listar_resenas(self, admin, busqueda=None):
        validar_es_administrador(admin)
        termino = validar_busqueda(busqueda)
        return self.resena_repo.listar_todas(termino)

    def eliminar_resena(self, admin, resena_id):
        validar_es_administrador(admin)
        self.resena_repo.eliminar(resena_id)
        return {'eliminado': True}

    # ── Resenas Multiples ─────────────────────────────────────────────────────

    def listar_resenas_multiples(self, admin, busqueda=None):
        validar_es_administrador(admin)
        termino = validar_busqueda(busqueda)
        return self.resena_multiple_repo.listar_todas(termino)

    def eliminar_resena_multiple(self, admin, resena_id):
        validar_es_administrador(admin)
        self.resena_multiple_repo.eliminar(resena_id)
        return {'eliminado': True}

    # ── Saldos Comerciales ────────────────────────────────────────────────────

    def listar_saldos(self, admin, busqueda=None):
        validar_es_administrador(admin)
        termino = validar_busqueda(busqueda)
        return self.saldo_repo.listar_todos(termino)

File: frontend/src/stores/adminPanel.js
/**
 * Pinia Store — Panel de Administracion (Sprint 2 HU3).
 *
 * Capa: stores/ (Estado del frontend)
 *
 * Gestiona el estado del panel admin: dashboard, listas de entidades,
 * seccion activa, termino de busqueda y operaciones CRUD.
 *
 * Flujo: AdminPanel.vue -> useAdminPanelStore -> AdminPanelRepository -> Backend
 */
import { defineStore } from 'pinia'
import { ref } from 'vue'
import AdminPanelRepository from '../services/api/AdminPanelApiService.js'

export const useAdminPanelStore = defineStore('adminPanel', () => {
  const api = new AdminPanelRepository()

  // Estado reactivo
  const dashboard = ref(null)
  const usuarios = ref([])
  const publicaciones = ref([])
  const trueques = ref([])
  const truequesMultiples = ref([])
  const resenas = ref([])
  const resenasMultiples = ref([])
  const saldos = ref([])
  const loading = ref(false)
  const error = ref('')

  // Dashboard
  async function cargarDashboard() {
    loading.value = true
    error.value = ''
    try {
      dashboard.value = await api.obtenerDashboard()
      return dashboard.value
    } catch (err) {
      error.value = err.message || 'Error al cargar dashboard.'
      throw err
    } finally {
      loading.value = false
    }
  }

  // Usuarios
  async function cargarUsuarios(busqueda = '') {
    loading.value = true
    error.value = ''
    try {
      const data = await api.obtenerUsuarios(busqueda)
      usuarios.value = data.usuarios || []
      return data
    } catch (err) {
      error.value = err.message || 'Error al cargar usuarios.'
      throw err
    } finally {
      loading.value = false
    }
  }

  async function toggleUsuario(usuarioId) {
    try {
      const data = await api.toggleUsuario(usuarioId)
      const idx = usuarios.value.findIndex((u) => u.id === usuarioId)
      if (idx !== -1 && data.usuario) usuarios.value[idx] = data.usuario
      return data
    } catch (err) {
      error.value = err.message || 'Error al cambiar estado del usuario.'
      throw err
    }
  }

  async function cambiarRol(usuarioId, isStaff) {
    try {
      const data = await api.cambiarRol(usuarioId, isStaff)
      const idx = usuarios.value.findIndex((u) => u.id === usuarioId)
      if (idx !== -1 && data.usuario) usuarios.value[idx] = data.usuario
      return data
    } catch (err) {
      error.value = err.message || 'Error al cambiar rol.'
      throw err
    }
  }

  async function eliminarUsuario(usuarioId) {
    try {
      const data = await api.eliminarUsuario(usuarioId)
      usuarios.value = usuarios.value.filter((u) => u.id !== usuarioId)
      return data
    } catch (err) {
      error.value = err.message || 'Error al eliminar usuario.'
      throw err
    }
  }

  async function editarUsuario(usuarioId, datos) {
    try {
      const data = await api.editarUsuario(usuarioId, datos)
      const idx = usuarios.value.findIndex((u) => u.id === usuarioId)
      if (idx !== -1 && data.usuario) usuarios.value[idx] = data.usuario
      return data
    } catch (err) {
      error.value = err.message || 'Error al editar usuario.'
      throw err
    }
  }

  // Publicaciones
  async function cargarPublicaciones(busqueda = '') {
    loading.value = true
    error.value = ''
    try {
      const data = await api.obtenerPublicaciones(busqueda)
      publicaciones.value = data.publicaciones || []
      return data
    } catch (err) {
      error.value = err.message || 'Error al cargar publicaciones.'
      throw err
    } finally {
      loading.value = false
    }
  }

  async function crearPublicacion(datos) {
    try {
      const data = await api.crearPublicacion(datos)
      if (data.publicacion) publicaciones.value.unshift(data.publicacion)
      return data
    } catch (err) {
      error.value = err.message || 'Error al crear publicacion.'
      throw err
    }
  }

  async function moderarPublicacion(publicacionId, estaActiva) {
    try {
      const data = await api.moderarPublicacion(publicacionId, estaActiva)
      const idx = publicaciones.value.findIndex((p) => p.id === publicacionId)
      if (idx !== -1 && data.publicacion) publicaciones.value[idx] = data.publicacion
      return data
    } catch (err) {
      error.value = err.message || 'Error al moderar publicacion.'
      throw err
    }
  }

  async function eliminarPublicacion(publicacionId) {
    try {
      const data = await api.eliminarPublicacion(publicacionId)
      publicaciones.value = publicaciones.value.filter((p) => p.id !== publicacionId)
      return data
    } catch (err) {
      error.value = err.message || 'Error al eliminar publicacion.'
      throw err
    }
  }

  async function editarPublicacion(publicacionId, datos) {
    try {
      const data = await api.editarPublicacion(publicacionId, datos)
      const idx = publicaciones.value.findIndex((p) => p.id === publicacionId)
      if (idx !== -1 && data.publicacion) publicaciones.value[idx] = data.publicacion
      return data
    } catch (err) {
      error.value = err.message || 'Error al editar publicacion.'
      throw err
    }
  }

  // Trueques
  async function cargarTrueques(busqueda = '') {
    loading.value = true
    error.value = ''
    try {
      const data = await api.obtenerTrueques(busqueda)
      trueques.value = data.trueques || []
      return data
    } catch (err) {
      error.value = err.message || 'Error al cargar trueques.'
      throw err
    } finally {
      loading.value = false
    }
  }

  async function actualizarEstadoTrueque(truequeId, estado) {
    try {
      const data = await api.actualizarEstadoTrueque(truequeId, estado)
      const idx = trueques.value.findIndex((t) => t.id === truequeId)
      if (idx !== -1 && data.trueque) trueques.value[idx] = data.trueque
      return data
    } catch (err) {
      error.value = err.message || 'Error al actualizar estado.'
      throw err
    }
  }

  async function eliminarTrueque(truequeId) {
    try {
      const data = await api.eliminarTrueque(truequeId)
      trueques.value = trueques.value.filter((t) => t.id !== truequeId)
      return data
    } catch (err) {
      error.value = err.message || 'Error al eliminar trueque.'
      throw err
    }
  }

  // Trueques Multiples
  async function cargarTruequesMultiples(busqueda = '') {
    loading.value = true
    error.value = ''
    try {
      const data = await api.obtenerTruequesMultiples(busqueda)
      truequesMultiples.value = data.trueques_multiples || []
      return data
    } catch (err) {
      error.value = err.message || 'Error al cargar trueques multiples.'
      throw err
    } finally {
      loading.value = false
    }
  }

  async function actualizarEstadoTruequeMultiple(truequeId, estado) {
    try {
      const data = await api.actualizarEstadoTruequeMultiple(truequeId, estado)
      const idx = truequesMultiples.value.findIndex((t) => t.id === truequeId)
      if (idx !== -1 && data.trueque_multiple) truequesMultiples.value[idx] = data.trueque_multiple
      return data
    } catch (err) {
      error.value = err.message || 'Error al actualizar estado.'
      throw err
    }
  }

  async function eliminarTruequeMultiple(truequeId) {
    try {
      const data = await api.eliminarTruequeMultiple(truequeId)
      truequesMultiples.value = truequesMultiples.value.filter((t) => t.id !== truequeId)
      return data
    } catch (err) {
      error.value = err.message || 'Error al eliminar trueque multiple.'
      throw err
    }
  }

  // Resenas
  async function cargarResenas(busqueda = '') {
    loading.value = true
    error.value = ''
    try {
      const data = await api.obtenerResenas(busqueda)
      resenas.value = data.resenas || []
      return data
    } catch (err) {
      error.value = err.message || 'Error al cargar resenas.'
      throw err
    } finally {
      loading.value = false
    }
  }

  async function eliminarResena(resenaId) {
    try {
      const data = await api.eliminarResena(resenaId)
      resenas.value = resenas.value.filter((r) => r.id !== resenaId)
      return data
    } catch (err) {
      error.value = err.message || 'Error al eliminar resena.'
      throw err
    }
  }

  // Resenas Multiples
  async function cargarResenasMultiples(busqueda = '') {
    loading.value = true
    error.value = ''
    try {
      const data = await api.obtenerResenasMultiples(busqueda)
      resenasMultiples.value = data.resenas_multiples || []
      return data
    } catch (err) {
      error.value = err.message || 'Error al cargar resenas multiples.'
      throw err
    } finally {
      loading.value = false
    }
  }

  async function eliminarResenaMultiple(resenaId) {
    try {
      const data = await api.eliminarResenaMultiple(resenaId)
      resenasMultiples.value = resenasMultiples.value.filter((r) => r.id !== resenaId)
      return data
    } catch (err) {
      error.value = err.message || 'Error al eliminar resena multiple.'
      throw err
    }
  }

  // Saldos
  async function cargarSaldos(busqueda = '') {
    loading.value = true
    error.value = ''
    try {
      const data = await api.obtenerSaldos(busqueda)
      saldos.value = data.saldos || []
      return data
    } catch (err) {
      error.value = err.message || 'Error al cargar saldos.'
      throw err
    } finally {
      loading.value = false
    }
  }

  return {
    dashboard, usuarios, publicaciones, trueques, truequesMultiples,
    resenas, resenasMultiples, saldos, loading, error,
    cargarDashboard, cargarUsuarios, toggleUsuario, cambiarRol, eliminarUsuario, editarUsuario,
    cargarPublicaciones, crearPublicacion, moderarPublicacion, eliminarPublicacion, editarPublicacion,
    cargarTrueques, actualizarEstadoTrueque, eliminarTrueque,
    cargarTruequesMultiples, actualizarEstadoTruequeMultiple, eliminarTruequeMultiple,
    cargarResenas, eliminarResena,
    cargarResenasMultiples, eliminarResenaMultiple,
    cargarSaldos,
  }
})
